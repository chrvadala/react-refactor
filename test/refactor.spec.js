const path = require("path");
const fs = require("fs")
const {removeSpaces} = require("./testUtils");
const {patchString} = require("../src/stringUtils");
const {execRefactor} = require("../src/refactor");
const {parseFile} = require('../src/parser')
const fixture = filename => path.join(__dirname, '__fixtures__', filename)

let example, refactoredExample;

beforeAll(done => {
  fs.readFile(fixture('example.jsx.txt'), 'utf8', (err, data) => {
    if (err) throw new Error()
    example = data
    done()
  })
})

beforeAll(done => {
  fs.readFile(fixture('refactoredExample.jsx.txt'), 'utf8', (err, data) => {
    if (err) throw new Error()
    refactoredExample = data
    done()
  })
})


describe('refactor', () => {
  it.only('should convert', () => {
    let parsedExample = parseFile(fixture('example.jsx.txt'))
    let {patch, skipped} = execRefactor(example, parsedExample)

    expect(Array.isArray(patch)).toBe(true)
    expect(Array.isArray(skipped)).toBe(true)
    expect(patch).toMatchSnapshot()
    expect(skipped).toMatchSnapshot()

    let output = patchString(example, patch)
    expect(removeSpaces(output))
      .toBe(removeSpaces(refactoredExample))
  })
})
