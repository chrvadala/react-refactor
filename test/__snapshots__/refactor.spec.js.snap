// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should parse Class file 1`] = `
Node {
  "body": Array [
    Node {
      "end": 26,
      "loc": SourceLocation {
        "end": Position {
          "column": 26,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "source": Node {
        "end": 25,
        "extra": Object {
          "raw": "'react'",
          "rawValue": "react",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 25,
            "line": 1,
          },
          "start": Position {
            "column": 18,
            "line": 1,
          },
        },
        "start": 18,
        "type": "StringLiteral",
        "value": "react",
      },
      "specifiers": Array [
        Node {
          "end": 12,
          "loc": SourceLocation {
            "end": Position {
              "column": 12,
              "line": 1,
            },
            "start": Position {
              "column": 7,
              "line": 1,
            },
          },
          "local": Node {
            "end": 12,
            "loc": SourceLocation {
              "end": Position {
                "column": 12,
                "line": 1,
              },
              "identifierName": "React",
              "start": Position {
                "column": 7,
                "line": 1,
              },
            },
            "name": "React",
            "start": 7,
            "type": "Identifier",
          },
          "start": 7,
          "type": "ImportDefaultSpecifier",
        },
      ],
      "start": 0,
      "type": "ImportDeclaration",
    },
    Node {
      "end": 62,
      "loc": SourceLocation {
        "end": Position {
          "column": 35,
          "line": 2,
        },
        "start": Position {
          "column": 0,
          "line": 2,
        },
      },
      "source": Node {
        "end": 61,
        "extra": Object {
          "raw": "'prop-types'",
          "rawValue": "prop-types",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 34,
            "line": 2,
          },
          "start": Position {
            "column": 22,
            "line": 2,
          },
        },
        "start": 49,
        "type": "StringLiteral",
        "value": "prop-types",
      },
      "specifiers": Array [
        Node {
          "end": 43,
          "loc": SourceLocation {
            "end": Position {
              "column": 16,
              "line": 2,
            },
            "start": Position {
              "column": 7,
              "line": 2,
            },
          },
          "local": Node {
            "end": 43,
            "loc": SourceLocation {
              "end": Position {
                "column": 16,
                "line": 2,
              },
              "identifierName": "PropTypes",
              "start": Position {
                "column": 7,
                "line": 2,
              },
            },
            "name": "PropTypes",
            "start": 34,
            "type": "Identifier",
          },
          "start": 34,
          "type": "ImportDefaultSpecifier",
        },
      ],
      "start": 27,
      "type": "ImportDeclaration",
    },
    Node {
      "body": Node {
        "body": Array [
          Node {
            "async": false,
            "body": Node {
              "body": Array [
                Node {
                  "end": 146,
                  "expression": Node {
                    "arguments": Array [
                      Node {
                        "end": 144,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 15,
                            "line": 6,
                          },
                          "identifierName": "props",
                          "start": Position {
                            "column": 10,
                            "line": 6,
                          },
                        },
                        "name": "props",
                        "start": 139,
                        "type": "Identifier",
                      },
                    ],
                    "callee": Node {
                      "end": 138,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 9,
                          "line": 6,
                        },
                        "start": Position {
                          "column": 4,
                          "line": 6,
                        },
                      },
                      "start": 133,
                      "type": "Super",
                    },
                    "end": 145,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 16,
                        "line": 6,
                      },
                      "start": Position {
                        "column": 4,
                        "line": 6,
                      },
                    },
                    "start": 133,
                    "type": "CallExpression",
                  },
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 17,
                      "line": 6,
                    },
                    "start": Position {
                      "column": 4,
                      "line": 6,
                    },
                  },
                  "start": 133,
                  "type": "ExpressionStatement",
                },
              ],
              "directives": Array [],
              "end": 150,
              "loc": SourceLocation {
                "end": Position {
                  "column": 3,
                  "line": 7,
                },
                "start": Position {
                  "column": 21,
                  "line": 5,
                },
              },
              "start": 127,
              "type": "BlockStatement",
            },
            "computed": false,
            "end": 150,
            "expression": false,
            "generator": false,
            "id": null,
            "key": Node {
              "end": 119,
              "loc": SourceLocation {
                "end": Position {
                  "column": 13,
                  "line": 5,
                },
                "identifierName": "constructor",
                "start": Position {
                  "column": 2,
                  "line": 5,
                },
              },
              "name": "constructor",
              "start": 108,
              "type": "Identifier",
            },
            "kind": "constructor",
            "loc": SourceLocation {
              "end": Position {
                "column": 3,
                "line": 7,
              },
              "start": Position {
                "column": 2,
                "line": 5,
              },
            },
            "params": Array [
              Node {
                "end": 125,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 19,
                    "line": 5,
                  },
                  "identifierName": "props",
                  "start": Position {
                    "column": 14,
                    "line": 5,
                  },
                },
                "name": "props",
                "start": 120,
                "type": "Identifier",
              },
            ],
            "start": 108,
            "static": false,
            "type": "ClassMethod",
          },
          Node {
            "async": false,
            "body": Node {
              "body": Array [
                Node {
                  "argument": Node {
                    "children": Array [
                      Node {
                        "end": 197,
                        "extra": null,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 6,
                            "line": 13,
                          },
                          "start": Position {
                            "column": 11,
                            "line": 11,
                          },
                        },
                        "start": 189,
                        "type": "JSXText",
                        "value": "

      ",
                      },
                    ],
                    "closingElement": Node {
                      "end": 203,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 12,
                          "line": 13,
                        },
                        "start": Position {
                          "column": 6,
                          "line": 13,
                        },
                      },
                      "name": Node {
                        "end": 202,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 11,
                            "line": 13,
                          },
                          "start": Position {
                            "column": 8,
                            "line": 13,
                          },
                        },
                        "name": "div",
                        "start": 199,
                        "type": "JSXIdentifier",
                      },
                      "start": 197,
                      "type": "JSXClosingElement",
                    },
                    "end": 203,
                    "extra": Object {
                      "parenStart": 176,
                      "parenthesized": true,
                    },
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 12,
                        "line": 13,
                      },
                      "start": Position {
                        "column": 6,
                        "line": 11,
                      },
                    },
                    "openingElement": Node {
                      "attributes": Array [],
                      "end": 189,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 11,
                          "line": 11,
                        },
                        "start": Position {
                          "column": 6,
                          "line": 11,
                        },
                      },
                      "name": Node {
                        "end": 188,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 10,
                            "line": 11,
                          },
                          "start": Position {
                            "column": 7,
                            "line": 11,
                          },
                        },
                        "name": "div",
                        "start": 185,
                        "type": "JSXIdentifier",
                      },
                      "selfClosing": false,
                      "start": 184,
                      "type": "JSXOpeningElement",
                    },
                    "start": 184,
                    "type": "JSXElement",
                  },
                  "end": 210,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 6,
                      "line": 14,
                    },
                    "start": Position {
                      "column": 4,
                      "line": 10,
                    },
                  },
                  "start": 169,
                  "type": "ReturnStatement",
                },
              ],
              "directives": Array [],
              "end": 214,
              "loc": SourceLocation {
                "end": Position {
                  "column": 3,
                  "line": 15,
                },
                "start": Position {
                  "column": 11,
                  "line": 9,
                },
              },
              "start": 163,
              "type": "BlockStatement",
            },
            "computed": false,
            "end": 214,
            "expression": false,
            "generator": false,
            "id": null,
            "key": Node {
              "end": 160,
              "loc": SourceLocation {
                "end": Position {
                  "column": 8,
                  "line": 9,
                },
                "identifierName": "render",
                "start": Position {
                  "column": 2,
                  "line": 9,
                },
              },
              "name": "render",
              "start": 154,
              "type": "Identifier",
            },
            "kind": "method",
            "loc": SourceLocation {
              "end": Position {
                "column": 3,
                "line": 15,
              },
              "start": Position {
                "column": 2,
                "line": 9,
              },
            },
            "params": Array [],
            "start": 154,
            "static": false,
            "type": "ClassMethod",
          },
        ],
        "end": 216,
        "loc": SourceLocation {
          "end": Position {
            "column": 1,
            "line": 16,
          },
          "start": Position {
            "column": 40,
            "line": 4,
          },
        },
        "start": 104,
        "type": "ClassBody",
      },
      "end": 216,
      "id": Node {
        "end": 79,
        "loc": SourceLocation {
          "end": Position {
            "column": 15,
            "line": 4,
          },
          "identifierName": "ClassComp",
          "start": Position {
            "column": 6,
            "line": 4,
          },
        },
        "name": "ClassComp",
        "start": 70,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 1,
          "line": 16,
        },
        "start": Position {
          "column": 0,
          "line": 4,
        },
      },
      "start": 64,
      "superClass": Node {
        "computed": false,
        "end": 103,
        "loc": SourceLocation {
          "end": Position {
            "column": 39,
            "line": 4,
          },
          "start": Position {
            "column": 24,
            "line": 4,
          },
        },
        "object": Node {
          "end": 93,
          "loc": SourceLocation {
            "end": Position {
              "column": 29,
              "line": 4,
            },
            "identifierName": "React",
            "start": Position {
              "column": 24,
              "line": 4,
            },
          },
          "name": "React",
          "start": 88,
          "type": "Identifier",
        },
        "property": Node {
          "end": 103,
          "loc": SourceLocation {
            "end": Position {
              "column": 39,
              "line": 4,
            },
            "identifierName": "Component",
            "start": Position {
              "column": 30,
              "line": 4,
            },
          },
          "name": "Component",
          "start": 94,
          "type": "Identifier",
        },
        "start": 88,
        "type": "MemberExpression",
      },
      "type": "ClassDeclaration",
    },
    Node {
      "end": 243,
      "expression": Node {
        "end": 242,
        "left": Node {
          "computed": false,
          "end": 237,
          "loc": SourceLocation {
            "end": Position {
              "column": 19,
              "line": 18,
            },
            "start": Position {
              "column": 0,
              "line": 18,
            },
          },
          "object": Node {
            "end": 227,
            "loc": SourceLocation {
              "end": Position {
                "column": 9,
                "line": 18,
              },
              "identifierName": "ClassComp",
              "start": Position {
                "column": 0,
                "line": 18,
              },
            },
            "name": "ClassComp",
            "start": 218,
            "type": "Identifier",
          },
          "property": Node {
            "end": 237,
            "loc": SourceLocation {
              "end": Position {
                "column": 19,
                "line": 18,
              },
              "identifierName": "propTypes",
              "start": Position {
                "column": 10,
                "line": 18,
              },
            },
            "name": "propTypes",
            "start": 228,
            "type": "Identifier",
          },
          "start": 218,
          "type": "MemberExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 24,
            "line": 18,
          },
          "start": Position {
            "column": 0,
            "line": 18,
          },
        },
        "operator": "=",
        "right": Node {
          "end": 242,
          "loc": SourceLocation {
            "end": Position {
              "column": 24,
              "line": 18,
            },
            "start": Position {
              "column": 22,
              "line": 18,
            },
          },
          "properties": Array [],
          "start": 240,
          "type": "ObjectExpression",
        },
        "start": 218,
        "type": "AssignmentExpression",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 25,
          "line": 18,
        },
        "start": Position {
          "column": 0,
          "line": 18,
        },
      },
      "start": 218,
      "type": "ExpressionStatement",
    },
    Node {
      "end": 272,
      "expression": Node {
        "end": 271,
        "left": Node {
          "computed": false,
          "end": 266,
          "loc": SourceLocation {
            "end": Position {
              "column": 22,
              "line": 19,
            },
            "start": Position {
              "column": 0,
              "line": 19,
            },
          },
          "object": Node {
            "end": 253,
            "loc": SourceLocation {
              "end": Position {
                "column": 9,
                "line": 19,
              },
              "identifierName": "ClassComp",
              "start": Position {
                "column": 0,
                "line": 19,
              },
            },
            "name": "ClassComp",
            "start": 244,
            "type": "Identifier",
          },
          "property": Node {
            "end": 266,
            "loc": SourceLocation {
              "end": Position {
                "column": 22,
                "line": 19,
              },
              "identifierName": "defaultProps",
              "start": Position {
                "column": 10,
                "line": 19,
              },
            },
            "name": "defaultProps",
            "start": 254,
            "type": "Identifier",
          },
          "start": 244,
          "type": "MemberExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 27,
            "line": 19,
          },
          "start": Position {
            "column": 0,
            "line": 19,
          },
        },
        "operator": "=",
        "right": Node {
          "end": 271,
          "loc": SourceLocation {
            "end": Position {
              "column": 27,
              "line": 19,
            },
            "start": Position {
              "column": 25,
              "line": 19,
            },
          },
          "properties": Array [],
          "start": 269,
          "type": "ObjectExpression",
        },
        "start": 244,
        "type": "AssignmentExpression",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 28,
          "line": 19,
        },
        "start": Position {
          "column": 0,
          "line": 19,
        },
      },
      "start": 244,
      "type": "ExpressionStatement",
    },
    Node {
      "declaration": Node {
        "end": 299,
        "loc": SourceLocation {
          "end": Position {
            "column": 24,
            "line": 22,
          },
          "identifierName": "ClassComp",
          "start": Position {
            "column": 15,
            "line": 22,
          },
        },
        "name": "ClassComp",
        "start": 290,
        "type": "Identifier",
      },
      "end": 300,
      "loc": SourceLocation {
        "end": Position {
          "column": 25,
          "line": 22,
        },
        "start": Position {
          "column": 0,
          "line": 22,
        },
      },
      "start": 275,
      "type": "ExportDefaultDeclaration",
    },
  ],
  "directives": Array [],
  "end": 301,
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 23,
    },
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "sourceType": "module",
  "start": 0,
  "type": "Program",
}
`;

exports[`parser should parse Func file 1`] = `
Node {
  "body": Array [
    Node {
      "end": 26,
      "loc": SourceLocation {
        "end": Position {
          "column": 26,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "source": Node {
        "end": 25,
        "extra": Object {
          "raw": "'react'",
          "rawValue": "react",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 25,
            "line": 1,
          },
          "start": Position {
            "column": 18,
            "line": 1,
          },
        },
        "start": 18,
        "type": "StringLiteral",
        "value": "react",
      },
      "specifiers": Array [
        Node {
          "end": 12,
          "loc": SourceLocation {
            "end": Position {
              "column": 12,
              "line": 1,
            },
            "start": Position {
              "column": 7,
              "line": 1,
            },
          },
          "local": Node {
            "end": 12,
            "loc": SourceLocation {
              "end": Position {
                "column": 12,
                "line": 1,
              },
              "identifierName": "React",
              "start": Position {
                "column": 7,
                "line": 1,
              },
            },
            "name": "React",
            "start": 7,
            "type": "Identifier",
          },
          "start": 7,
          "type": "ImportDefaultSpecifier",
        },
      ],
      "start": 0,
      "type": "ImportDeclaration",
    },
    Node {
      "end": 62,
      "loc": SourceLocation {
        "end": Position {
          "column": 35,
          "line": 2,
        },
        "start": Position {
          "column": 0,
          "line": 2,
        },
      },
      "source": Node {
        "end": 61,
        "extra": Object {
          "raw": "'prop-types'",
          "rawValue": "prop-types",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 34,
            "line": 2,
          },
          "start": Position {
            "column": 22,
            "line": 2,
          },
        },
        "start": 49,
        "type": "StringLiteral",
        "value": "prop-types",
      },
      "specifiers": Array [
        Node {
          "end": 43,
          "loc": SourceLocation {
            "end": Position {
              "column": 16,
              "line": 2,
            },
            "start": Position {
              "column": 7,
              "line": 2,
            },
          },
          "local": Node {
            "end": 43,
            "loc": SourceLocation {
              "end": Position {
                "column": 16,
                "line": 2,
              },
              "identifierName": "PropTypes",
              "start": Position {
                "column": 7,
                "line": 2,
              },
            },
            "name": "PropTypes",
            "start": 34,
            "type": "Identifier",
          },
          "start": 34,
          "type": "ImportDefaultSpecifier",
        },
      ],
      "start": 27,
      "type": "ImportDeclaration",
    },
    Node {
      "async": false,
      "body": Node {
        "body": Array [
          Node {
            "argument": Node {
              "children": Array [
                Node {
                  "end": 117,
                  "extra": null,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 4,
                      "line": 8,
                    },
                    "start": Position {
                      "column": 9,
                      "line": 6,
                    },
                  },
                  "start": 111,
                  "type": "JSXText",
                  "value": "

    ",
                },
              ],
              "closingElement": Node {
                "end": 123,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 10,
                    "line": 8,
                  },
                  "start": Position {
                    "column": 4,
                    "line": 8,
                  },
                },
                "name": Node {
                  "end": 122,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 9,
                      "line": 8,
                    },
                    "start": Position {
                      "column": 6,
                      "line": 8,
                    },
                  },
                  "name": "div",
                  "start": 119,
                  "type": "JSXIdentifier",
                },
                "start": 117,
                "type": "JSXClosingElement",
              },
              "end": 123,
              "extra": Object {
                "parenStart": 100,
                "parenthesized": true,
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 10,
                  "line": 8,
                },
                "start": Position {
                  "column": 4,
                  "line": 6,
                },
              },
              "openingElement": Node {
                "attributes": Array [],
                "end": 111,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 9,
                    "line": 6,
                  },
                  "start": Position {
                    "column": 4,
                    "line": 6,
                  },
                },
                "name": Node {
                  "end": 110,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 8,
                      "line": 6,
                    },
                    "start": Position {
                      "column": 5,
                      "line": 6,
                    },
                  },
                  "name": "div",
                  "start": 107,
                  "type": "JSXIdentifier",
                },
                "selfClosing": false,
                "start": 106,
                "type": "JSXOpeningElement",
              },
              "start": 106,
              "type": "JSXElement",
            },
            "end": 128,
            "loc": SourceLocation {
              "end": Position {
                "column": 4,
                "line": 9,
              },
              "start": Position {
                "column": 2,
                "line": 5,
              },
            },
            "start": 93,
            "type": "ReturnStatement",
          },
        ],
        "directives": Array [],
        "end": 130,
        "loc": SourceLocation {
          "end": Position {
            "column": 1,
            "line": 10,
          },
          "start": Position {
            "column": 25,
            "line": 4,
          },
        },
        "start": 89,
        "type": "BlockStatement",
      },
      "end": 130,
      "expression": false,
      "generator": false,
      "id": Node {
        "end": 81,
        "loc": SourceLocation {
          "end": Position {
            "column": 17,
            "line": 4,
          },
          "identifierName": "FuncComp",
          "start": Position {
            "column": 9,
            "line": 4,
          },
        },
        "name": "FuncComp",
        "start": 73,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 1,
          "line": 10,
        },
        "start": Position {
          "column": 0,
          "line": 4,
        },
      },
      "params": Array [
        Node {
          "end": 87,
          "loc": SourceLocation {
            "end": Position {
              "column": 23,
              "line": 4,
            },
            "identifierName": "props",
            "start": Position {
              "column": 18,
              "line": 4,
            },
          },
          "name": "props",
          "start": 82,
          "type": "Identifier",
        },
      ],
      "start": 64,
      "type": "FunctionDeclaration",
    },
    Node {
      "end": 156,
      "expression": Node {
        "end": 155,
        "left": Node {
          "computed": false,
          "end": 150,
          "loc": SourceLocation {
            "end": Position {
              "column": 18,
              "line": 12,
            },
            "start": Position {
              "column": 0,
              "line": 12,
            },
          },
          "object": Node {
            "end": 140,
            "loc": SourceLocation {
              "end": Position {
                "column": 8,
                "line": 12,
              },
              "identifierName": "FuncComp",
              "start": Position {
                "column": 0,
                "line": 12,
              },
            },
            "name": "FuncComp",
            "start": 132,
            "type": "Identifier",
          },
          "property": Node {
            "end": 150,
            "loc": SourceLocation {
              "end": Position {
                "column": 18,
                "line": 12,
              },
              "identifierName": "propTypes",
              "start": Position {
                "column": 9,
                "line": 12,
              },
            },
            "name": "propTypes",
            "start": 141,
            "type": "Identifier",
          },
          "start": 132,
          "type": "MemberExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 23,
            "line": 12,
          },
          "start": Position {
            "column": 0,
            "line": 12,
          },
        },
        "operator": "=",
        "right": Node {
          "end": 155,
          "loc": SourceLocation {
            "end": Position {
              "column": 23,
              "line": 12,
            },
            "start": Position {
              "column": 21,
              "line": 12,
            },
          },
          "properties": Array [],
          "start": 153,
          "type": "ObjectExpression",
        },
        "start": 132,
        "type": "AssignmentExpression",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 24,
          "line": 12,
        },
        "start": Position {
          "column": 0,
          "line": 12,
        },
      },
      "start": 132,
      "type": "ExpressionStatement",
    },
    Node {
      "end": 184,
      "expression": Node {
        "end": 183,
        "left": Node {
          "computed": false,
          "end": 178,
          "loc": SourceLocation {
            "end": Position {
              "column": 21,
              "line": 13,
            },
            "start": Position {
              "column": 0,
              "line": 13,
            },
          },
          "object": Node {
            "end": 165,
            "loc": SourceLocation {
              "end": Position {
                "column": 8,
                "line": 13,
              },
              "identifierName": "FuncComp",
              "start": Position {
                "column": 0,
                "line": 13,
              },
            },
            "name": "FuncComp",
            "start": 157,
            "type": "Identifier",
          },
          "property": Node {
            "end": 178,
            "loc": SourceLocation {
              "end": Position {
                "column": 21,
                "line": 13,
              },
              "identifierName": "defaultProps",
              "start": Position {
                "column": 9,
                "line": 13,
              },
            },
            "name": "defaultProps",
            "start": 166,
            "type": "Identifier",
          },
          "start": 157,
          "type": "MemberExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 26,
            "line": 13,
          },
          "start": Position {
            "column": 0,
            "line": 13,
          },
        },
        "operator": "=",
        "right": Node {
          "end": 183,
          "loc": SourceLocation {
            "end": Position {
              "column": 26,
              "line": 13,
            },
            "start": Position {
              "column": 24,
              "line": 13,
            },
          },
          "properties": Array [],
          "start": 181,
          "type": "ObjectExpression",
        },
        "start": 157,
        "type": "AssignmentExpression",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 27,
          "line": 13,
        },
        "start": Position {
          "column": 0,
          "line": 13,
        },
      },
      "start": 157,
      "type": "ExpressionStatement",
    },
    Node {
      "declaration": Node {
        "end": 210,
        "loc": SourceLocation {
          "end": Position {
            "column": 23,
            "line": 16,
          },
          "identifierName": "FuncComp",
          "start": Position {
            "column": 15,
            "line": 16,
          },
        },
        "name": "FuncComp",
        "start": 202,
        "type": "Identifier",
      },
      "end": 211,
      "loc": SourceLocation {
        "end": Position {
          "column": 24,
          "line": 16,
        },
        "start": Position {
          "column": 0,
          "line": 16,
        },
      },
      "start": 187,
      "type": "ExportDefaultDeclaration",
    },
  ],
  "directives": Array [],
  "end": 212,
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 17,
    },
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "sourceType": "module",
  "start": 0,
  "type": "Program",
}
`;
